File: luogu/haskell/src/P3367.hs
module P3367 where

import qualified Data.Vector.Unboxed as V
import Data.Vector.Unboxed ((!), (//))

type UnionFind = V.Vector Int

ask :: UnionFind -> Int -> (UnionFind, Int)
ask vec i = if vec ! i == i then (vec, i) else 
    let (vec', result) = ask vec $ vec ! i in
        (vec' // [(i, result)], result)

cat :: UnionFind -> Int -> Int -> UnionFind
cat vec x y = let (vec', x') = ask vec x in
    let (vec'', y') = ask vec' y in
        vec // [(x', y')]

resolve :: Int -> UnionFind -> IO ()
resolve 0 _ = return ()
resolve n uf = do
    [z, x, y] <- map read . words <$> getLine :: IO [Int]

    uf' <- if z == 1 then return $ cat uf x y else 
        let (vec, x') = ask uf x
            (vec', y') = ask vec y in 
            putStrLn (if x' == y' then "Y" else "N") >>= const (return vec')

    resolve (n - 1) uf'

main :: IO ()
main = do
    [n, m] <- map read . words <$> getLine :: IO [Int]

    resolve m $ V.generate (n + 1) id

    return ()

File: luogu/rust/src/p3367.rs
use std::io;

struct UnionFind {
    arr: Vec<usize>
}

impl UnionFind {
    fn new(len: usize) -> UnionFind {
        let mut vec: Vec<usize> = Vec::new();

        for i in 0..len {
            vec.push(i);
        };

        UnionFind {
            arr: vec
        }
    }

    fn ask(&mut self, x: usize) -> usize {
        if self.arr[x] == x {
            x
        } else {
            let p = self.arr[x];
            self.arr[x] = self.ask(p);
            self.arr[x]
        }
    }

    fn cat(&mut self, x: usize, y: usize) {
        let new_x = self.ask(x) as usize;

        self.arr[new_x] = self.ask(y);
    }
}

pub fn main() {
    let mut line = String::new();

    io::stdin().read_line(&mut line).expect("qaq");

    let words: Vec<&str> = line.trim().split(' ').collect();

    let len: usize = words[0].parse().expect("qwq");
    let req_count: usize = words[1].parse().expect("qwq");

    let mut uf = UnionFind::new(len + 1);

    for i in 0..req_count {
        let mut line = String::new();

        io::stdin().read_line(&mut line).expect("qaq");

        let words: Vec<&str> = line.trim().split(' ').collect();
        let req: i32 = words[0].parse().expect("qwq");
        let x: usize = words[1].parse().expect("qwq");
        let y: usize = words[2].parse().expect("qwq");

        if req == 1 {
            uf.cat(x, y);
        } else {
            let result_x = uf.ask(x);
            let result_y = uf.ask(y);

            if result_x == result_y {
                println!("Y")
            } else {
                println!("N")
            }
        }
    }
}